test/test_nn.py

```diff
@@ -6058,6 +6058,168 @@ def test_shape(N, C, ID, IH, IW, D, H, W, mode, padding_mode, align_corners):
                     for input_requires_grad in [False, True]:
                         test(N, C, D, H, W, mode, padding_mode, align_corners, input_requires_grad)
 
+    def test_grid_sample_nearest_neighbor_rounding_mode_consistency(self):
+
+        device_list = ['cpu']
+        if TEST_CUDA:
+            device_list.append('cuda')
+
+        def normalize_indices(indices_unnormalized: torch.Tensor, dim_size: int, align_corners: bool):
+            if align_corners:
+                indices_normalized = 2 * indices_unnormalized / (dim_size - 1) - 1
+            else:
+                indices_normalized = (indices_unnormalized * 2 + 1) / dim_size - 1
+            return indices_normalized
+
+        test_dim_size = 10
+        non_test_dim_size = 9
+        step_size = 0.1
+
+        batch_size = 1
+        channel_size = 1
+
+        mode = 'nearest'
+        for device in device_list:
+            for padding_mode in ('zeros', 'border', 'reflection'):
+                for align_corners in (True, False):
+                    # Normalized inquery indices
+                    inquery_indices_unnormalized = torch.arange(
+                        0,
+                        test_dim_size - 1 + step_size, step_size,
+                        dtype=torch.float32,
+                        device=device
+                    )
+                    # Note that even though we are trying to create normalized indices
+                    # which results in x.0 and x.5 indices after unnormalization,
+                    # because of the numerical error,
+                    # the rounding direction might not always be expected as designed.
+                    # The best we could do is to ensure the rounding behaviors across
```

Comment:
I was able to prepare the round-to-the-nearest-even ground truth for all the x.0 and x.5 values in my first implementation of the unit test which has not been uploaded. Basically, if you create an unnormalized array using literals, such as [0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0, ...], the rounding behaviors will be exactly the same as what the operator, either `std::round` or `std::nearbyint`, defined.

Then I used the above unnormalized array to create an normalized array for grid sample, a small numeric error may or may not happen. The grid sample implementation will further unnormalized the array and it will introduce additional numerical errors.

So if I try to compare the grid sample interpolation result with my expected rounding behavior ground truth, in practice, they cannot be bitwise identical. There would be 1/5 - 1/20 values that are different in the array from the tests I conducted.

Reply:
Makes sense, thanks for the clarification!